ansicolor.js

```diff
@@ -198,131 +213,220 @@ const TEXT    = 0,
       BRACKET = 1,
       CODE    = 2
 
-function rawParse (s) {
-    
-    let state = TEXT, buffer = '', text = '', code = '', codes = []
-    let spans = []
-
-    for (let i = 0, n = s.length; i < n; i++) {
-
-        const c = s[i]
+class Span {
+  constructor(code, text) {
+    this.code = code;
+    this.text = text;
+
+    // Those are added in the actual parse, this is done for performance reasons to have the same hidden class
+    this.css = "";
+    this.color = "";
+    this.bgColor = "";
+    this.bold = undefined;
+    this.inverse = undefined;
+    this.italic = undefined;
+    this.underline = undefined;
+    this.bright = undefined;
+    this.dim = undefined;
+  }
+}
 
-        buffer += c
+// getString as function instead of string to allow garbage collection
+function* rawParse(getString) {
```

Comment:
Because the processing can take time we don't want the user to not do anything with the data until all of it is ready, so this way the user can do something with it

and some use cases just don't need some of the spans, so it will be removed faster

consider this:
```javascript

function createSpans() {
	return [{text: 'hello'}, {text: ''}, {text: 'world'}];
}

function run() {
	const allSpans = createSpans()

	for(const span of spans) {
		if(span.text.length === 0) {
			continue; // We don't care for empty spans
		}
	}
}
```

in this example, the span that we don't care can be GCed **after** the `run` function finish

but if we use generators the empty span **can** be GCed immediately:

```javascript
function* createSpans() {
	yield {text: 'hello'};
	yield {text: ''};
	yield {text: 'world'};
}

function run() {
	const allSpans = createSpans()

	for(const span of spans) {
		if(span.text.length === 0) {
			continue; // We don't care for empty spans
		}
	}
}
```

(there is no reference to the empty span)